### PR TITLE
feat: hostRules rewrite

### PR DESCRIPTION
**Pull request by rarkins**

### CODE REVIEW

File: lib/datasource/docker/index.js [hunk 1]
```diff
@@ -50,28 +52,27 @@ async function getAuthHeaders(registry, repository) {
...
 
-    const { host } = URL.parse(registry);
-    const opts = hostRules.find({ hostType: 'docker', host }) || {};
-    opts.json = true;
-    if (opts.username && opts.password) {
-      const auth = Buffer.from(`${opts.username}:${opts.password}`).toString(
-        'base64'
-      );
-      opts.headers = { Authorization: `Basic ${auth}` };
+    const hostOpts = hostRules.find({ hostType, url: apiCheckUrl });
```
Comment (ViceIce) [inline lib/datasource/docker/index.js]: we should use ` { ...hostRules.find({ hostType, url: apiCheckUrl }), json: true }` to be sure hostOpts are not null

File: lib/datasource/npm/npmrc.js [hunk 2]
```diff
@@ -15,6 +15,7 @@ function getNpmrc() {
 }
 
 function setNpmrc(input) {
+  debugger;
```
Comment (ViceIce) [inline lib/datasource/npm/npmrc.js]: Ups, forbidden 😄

File: lib/datasource/packagist/index.js [hunk 3]
```diff
@@ -11,19 +11,20 @@ module.exports = {
...
-  const opts = {
-    ...hostRules.find({ hostType: 'packagist', host }),
-    json: true,
-  };
-  if (opts && opts.username && opts.password) {
-    const auth = Buffer.from(`${opts.username}:${opts.password}`).toString(
-      'base64'
-    );
-    opts.headers = { Authorization: `Basic ${auth}` };
+  const hostOpts = hostRules.find({ hostType, url });
```
Comment (ViceIce) [inline lib/datasource/packagist/index.js]: Use spread operator too

File: lib/manager/composer/artifacts.js [hunk 4]
```diff
@@ -39,50 +39,44 @@ async function getArtifacts(
...
       };
     }
     try {
       // istanbul ignore else
       if (is.array(config.registryUrls)) {
         for (const regUrl of config.registryUrls) {
           if (regUrl.url) {
-            const { host } = URL.parse(regUrl.url);
-            const hostRule = hostRules.find({
+            const hostOpts = hostRules.find({
```
Comment (ViceIce) [inline lib/manager/composer/artifacts.js]: here too

Comment (ViceIce) [reply lib/manager/composer/artifacts.js]: or null check for `hostOpts`

File: lib/manager/npm/post-update/index.js [hunk 5]
```diff
@@ -304,7 +304,10 @@ async function writeUpdatedPackageFiles(config) {
     delete massagedFile.engines;
     delete massagedFile.scripts;
     try {
-      const { token } = hostRules.find({ hostType: config.platform });
+      const { token } = hostRules.find({
```
Comment (ViceIce) [inline lib/manager/npm/post-update/index.js]: what happens when find returns `null`?

File: lib/manager/npm/post-update/index.js [hunk 6]
```diff
@@ -377,7 +380,10 @@ async function getAdditionalFiles(config, packageFiles) {
 
   let token = '';
   try {
-    ({ token } = hostRules.find({ hostType: config.platform }));
+    ({ token } = hostRules.find({
```
Comment (ViceIce) [inline lib/manager/npm/post-update/index.js]: or here? will spread throw an error?

**Review by viceice — CHANGES_REQUESTED** [review]
Maybe we should always return an empty hostrule instead or null in `hostRule.find` ?

File: lib/manager/npm/post-update/index.js (hunk 6 above)
Comment (ViceIce) [reply lib/manager/npm/post-update/index.js]: yes: 
```
"Cannot destructure property \`token\` of 'undefined' or 'null'."
```

**Comment by rarkins** [comment]
Q1: How should we "define" the special keys for `hostRules`?

- `hostType`: e.g. github, gitlab, npm, pypi, docker, etc. Optional, because someone might like to configure credentials that are shared for multiple services on the same host
- `hostName`: e.g. `registry.mycompany.com`. Optional, because `baseUrl` could be provided instead (btw I'm unsure whether to put this all lower case or not)
- `baseUrl`, e.g. `https://registry.mycompany.com`. Interchangeable with `hostName` in most cases (typical users won't have more than one credentials per host) but would even allow for a specific token for a certain repo on github for examples (`https://api.github.com/repos/someotherowner/`)

We should not allow both `hostName` and `baseUrl` in the same rule.

Q2: How should we *query* hostRules?

- In every case so far I've seen we have a hostType. i.e. we don't just query a URL for random things. But also I don't think we *have* to make it mandatory in case there's a future use I'm not thinking of
- Usually we'll have a full URL we are querying. In that case hostType must match each rule's hostType - unless the rule's hostType is null - and then match either the hostName or baseUrl prefix.

Q3: What should we return if no match?

- Currently we return `null` if no match. Should we return `{}` instead?. `{}` will make results easier to work with, but make it less convenient to know "do we have a result or not?". But I think in most cases we don't care if there's *any* result, we care if there's a token or username/password.

**Comment by rarkins** [comment]
@ViceIce I'm thinking that it makes the `.find()` interface easier if we always return an object rather than `null`. What do you think?

**Comment by ViceIce** [comment]
> Q1: How should we "define" the special keys for `hostRules`?
> 
We can add multiple interfaces / overloaded function definitions.
[ts function overloads](https://www.typescriptlang.org/docs/handbook/functions.html#overloads)

> We should not allow both `hostName` and `baseUrl` in the same rule.
> 
> Q2: How should we _query_ hostRules?
> 
We can make `hostType` optional in both overloads.
> 
> Q3: What should we return if no match?
> 
That's the question. I'm open to both solutions. I prefer `null`, but then we have to add a lot null checks or use the object spread operator.

**Comment by ViceIce** [comment]
> @ViceIce I'm thinking that it makes the `.find()` interface easier if we always return an object rather than `null`. What do you think?

So we can never check existance, but we can add another `exists` function if required.

**Comment by rarkins** [comment]
Q4: What should the return structure look like?

We have standardised on `got` and would like to expose some of its flexibility to the users (both bot admins and repo admins alike). Examples include for things like timeouts or retries, etc. But also there are some "rules" that are for our Renovate use, such as username/password that we don't pass through. I can think of some other users in future too, such as allowing users to specify if a datasource host is critical or not (i.e. when to throw `registry-failure` or not).

I'm therefore considering that we define a return interface with known/specified fields, that *includes* an `options` object which corresponds to got's options and we pass transparently through.

So a rule for Docker could look like this:
```json
{
  "hostType": "docker",
  "username": "abc",
  "password": "123",
  "options": {
    "json": true
  }
}
```

**Comment by rarkins** [comment]
I just realised that we need to support matching on partial domain too. e.g. we'd want a rule for `*.docker.io`. What would be the best syntax for that? This could potentially also help for `*.github.com`

**Comment by ViceIce** [comment]
> Q4: What should the return structure look like?
> 
Move got settings to an options property is a good idea.

**Comment by rarkins** [comment]
OK so let's return an object at all times to simplify the return handling. If there are cases where we need to ask "does this query match any rules?" then we can add a specific new interface for that.

**Comment by ViceIce** [comment]
Sound good

**Comment by rarkins** [comment]
Proposal: yet another hostRule key to be supported: `domainName`. i.e. to allow `docker.io` or `github.com`

**Comment by ViceIce** [comment]
Should be mutual exclusive to `hostName` and `baseUrl`? (at least for the `find` args)

**Comment by rarkins** [comment]
So:
- (a) You can have maximum one of hostName, domainName or baseUrl
- (a) You can optionally have hostType
- You must have at least one of the above, i.e. a or b or (a&b)

**Comment by rarkins** [comment]
Although.. what if we want to allow rules that apply to every request, such as overriding the headers.user-agent field?

**Comment by ViceIce** [comment]
> Although.. what if we want to allow rules that apply to every request, such as overriding the headers.user-agent field?

Good question. Use `{}` or `null` as query?

**Comment by rarkins** [comment]
> Good question. Use {} or null as query?

I wasn't thinking along those lines. More like people would want to push a rule like this that applies to every URL:

```json
{
  "options": {
    "headers": {
      "user-agent": "something"
    }
  }
}
```

So then when you query for any url or hostType, it will always include this header. Note: this would be future functionality, not day 1.

**Comment by rarkins** [comment]
Rules to then be applied in this order:
- No hostType/hostName/domainName/baseUrl
- hostType only
- domainName only
- hostName only
- baseUrl only
- hostType + any of the other three

**Comment by ViceIce** [comment]
Ok, sure. I've misunderstood. 🙈

**Pull request merged**